contracts/clients/CkbProof.sol

```diff
@@ -0,0 +1,365 @@
+        }
+        return header;
+    }
+}
+
+// Define ckb blake2b
+contract Blake2b {
+    function blake2b(bytes memory data) public view returns (bytes32) {
+        // axon_precompile_address(0x06)
+        address blake2b_addr = address(0x0106);
+        (bool isSuccess, bytes memory res) = blake2b_addr.staticcall(
+            abi.encode(data)
+        );
+
+        bytes32 hash;
+        if (isSuccess) {
+            hash = abi.decode(res, (bytes32));
+        }
+        return hash;
+    }
+}
+
+contract CkbMbt {
+    function verify(
+        VerifyProofPayload memory payload
+    ) public view returns (bool) {
+        // axon_precompile_address(0x07)
+        address ckb_mbt_addr = address(0x0107);
+        (bool isSuccess, bytes memory res) = ckb_mbt_addr.staticcall(
+            abi.encode(payload)
+        );
+
+        uint8[] memory verify_success;
+        if (isSuccess) {
+            verify_success = abi.decode(res, (uint8[]));
+        } else {
+            return false;
+        }
+        return verify_success[0] == 1;
+    }
+}
+
+function decodeProof(
+    RLPReader.RLPItem[] memory items
+) pure returns (Proof memory) {
+    Proof memory proof;
+
+    // Decode indices
+    RLPReader.RLPItem[] memory indicesItems = RLPReader.toList(items[0]);
+    // proof.indices = new uint32;
+    for (uint i = 0; i < indicesItems.length; i++) {
+        proof.indices[i] = uint32(RLPReader.toUint(indicesItems[i]));
+    }
+
+    // Decode lemmas
+    RLPReader.RLPItem[] memory lemmasItems = RLPReader.toList(items[1]);
+    // proof.lemmas = new bytes32;
+    for (uint i = 0; i < lemmasItems.length; i++) {
+        proof.lemmas[i] = bytes32(RLPReader.toBytes(lemmasItems[i]));
+    }
+
+    // Decode leaves
+    RLPReader.RLPItem[] memory leavesItems = RLPReader.toList(items[2]);
+    // proof.leaves = new bytes32;
+    for (uint i = 0; i < leavesItems.length; i++) {
+        proof.leaves[i] = bytes32(RLPReader.toBytes(leavesItems[i]));
+    }
+
+    return proof;
+}
+
+function parseProof(
+    bytes calldata abiEncodedProof
+) pure returns (AxonObjectProof memory) {
+    AxonObjectProof memory proof = abi.decode(
+        abiEncodedProof,
+        (AxonObjectProof)
+    );
+    return proof;
+}
+
+function calculateHashes(
+    bytes memory ckbTransaction
+) view returns (bytes32 transactionHash, bytes32 witnessHash) {
+    // Calculate the hashes here
+    (uint256 offset, uint256 size) = ckbTransaction.readCKBTxRaw();
+    bytes memory raw_tx = new bytes(size);
+    for (uint i = 0; i < size; i++) {
+        raw_tx[i] = ckbTransaction[i + offset];
+    }
+
+    Blake2b blake2b;
+    return (blake2b.blake2b(raw_tx), blake2b.blake2b(ckbTransaction));
+}
+
+function decodeRlpEnvelope(
+    bytes memory rlpEncodedData
+) pure returns (Envelope memory) {
+    RLPReader.RLPItem[] memory ls = rlpEncodedData.toRlpItem().toList();
+
+    // Decode the msg_type
+    MsgType msg_type = MsgType(ls[0].toUint());
```

Comment:
Will revert because `MsgType` declaration is incomplete. Maybe just ignore the `msg_type`.